### PR TITLE
[PTSQ 2/4][Torch] Nested and successive hooks

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
Part 2/4 of SQ support https://github.com/openvinotoolkit/nncf/pull/2279

### Changes

* Nested and successive hooks are supported

### Reason for changes

* To make it possible to insert nested and successive hooks to the torch models

### Related tickets

124563

### Tests

* tests/torch/test_model_transformer.py checks sequential and nested hooks are inserted correctly / checks temporary hooks are applied correctly

### CODE REVIEW

File: nncf/torch/nncf_network.py [hunk 1]
```diff
@@ -716,13 +718,15 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
         return result
 
     def get_node_to_op_address_mapping(self) -> Dict[NNCFNodeName, OperationAddress]:
-        # The IDs of corresponding nodes of the original dynamic graph and original NNCF graph
```
Comment (AlexanderDokuchaev) [inline nncf/torch/nncf_network.py:719]: Add docstring and typehint for return value

File: nncf/torch/nncf_network.py [hunk 2]
```diff
@@ -716,13 +718,15 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
...
-        # must be equal for this to work.
         retval = {}
-        for node in self._original_dynamic_graph.get_all_nodes():
+        dynamic_graph = (
+            self._original_dynamic_graph if self._compressed_traced_graph is None else self._compressed_traced_graph
+        )
+        nncf_graph = self._original_graph if self._compressed_graph is None else self._compressed_graph
```
Comment (AlexanderDokuchaev) [inline nncf/torch/nncf_network.py:725]: is it possible that `self._compressed_traced_graph` is exists but `self._compressed_graph`? 


```python
if self._compressed_graph is None:
   dynamic_graph = self._original_dynamic_graph
   nncf_graph = self._original_graph
else: 
   dynamic_graph = self._compressed_traced_graph
   nncf_graph = self._compressed_graph
```

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/torch/nncf_network.py (hunk 2 above)
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py:725]: It is not possible in current code, but I would prefer to check both graphs explicitly as dynamic graph and NNCFGraph are not explicitly depend on each other. I would rather avoid introducing of this implicit dependency

File: nncf/torch/dynamic_graph/context.py [hunk 3]
```diff
@@ -91,8 +92,8 @@ class TracingContext:
     def __init__(self):
         self.graph = DynamicGraph()
 
-        self._post_hooks = {}
-        self._pre_hooks: Dict[PreHookId, List[Callable]] = {}
+        self._post_hooks = defaultdict(list)
+        self._pre_hooks: Dict[PreHookId, List[Callable]] = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/dynamic_graph/context.py]: Done, thanks

File: tests/common/test_statistics_aggregator.py [hunk 4]
```diff
@@ -73,6 +74,7 @@ def get_statistics_aggregator(self, dataset):
     def get_dataset(self, samples):
         pass
 
+    @staticmethod
     @abstractmethod
     def get_target_point(self, target_type: TargetType) -> TargetPoint:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/common/test_statistics_aggregator.py]: Done

File: tests/openvino/native/test_statistics_aggregator.py [hunk 5]
```diff
@@ -50,7 +50,8 @@ def get_StatisticAgregatorTestModel(input_shape, kernel):
 
 
 class TestStatisticsAggregator(TemplateTestStatisticsAggregator):
-    def get_min_max_algo_backend_cls(self) -> Type[OVMinMaxAlgoBackend]:
+    @classmethod
+    def get_min_max_algo_backend_cls(cls) -> Type[OVMinMaxAlgoBackend]:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/openvino/native/test_statistics_aggregator.py]: Done

File: nncf/torch/nncf_network.py (hunk 1 above)
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py:719]: Done

File: nncf/torch/nncf_network.py (hunk 2 above)
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: I made a `PTGraphPair` to make it clear, please check

**Review by AlexanderDokuchaev — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
LGTM

**Pull request merged**